pkg/blockbuilder/scheduler/jobs.go

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package scheduler
+
+import (
+	"container/heap"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+)
+
+var (
+	errNoJobAvailable = errors.New("no job available")
+	errJobNotFound    = errors.New("job not found")
+	errJobNotAssigned = errors.New("job not assigned to worker")
+)
+
+type jobQueue struct {
+	leaseTime time.Duration
+	logger    log.Logger
+
+	mu         sync.Mutex
+	jobs       map[string]*job
+	unassigned jobHeap
+}
+
+func newJobQueue(leaseTime time.Duration, logger log.Logger) *jobQueue {
+	return &jobQueue{
+		leaseTime: leaseTime,
+		logger:    logger,
+
+		jobs: make(map[string]*job),
+	}
+}
+
+// assign assigns the highest-priority unassigned job to the given worker.
+func (s *jobQueue) assign(workerID string) (string, jobSpec, error) {
+	if workerID == "" {
+		return "", jobSpec{}, errors.New("workerID cannot not be empty")
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.unassigned.Len() == 0 {
+		return "", jobSpec{}, errNoJobAvailable
+	}
+
+	j := heap.Pop(&s.unassigned).(*job)
+	j.assignee = workerID
+	j.leaseExpiry = time.Now().Add(s.leaseTime)
+	return j.id, j.spec, nil
+}
+
+// addOrUpdate adds a new job or updates an existing job with the given spec.
+func (s *jobQueue) addOrUpdate(id string, spec jobSpec) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if j, ok := s.jobs[id]; ok {
+		// We can only update an unassigned job.
+		if j.assignee == "" {
+			j.spec = spec
+		}
+	} else {
+		j = &job{
+			id:          id,
+			assignee:    "",
+			leaseExpiry: time.Now().Add(s.leaseTime),
+			failCount:   0,
+			spec:        spec,
+		}
+		s.jobs[id] = j
+		heap.Push(&s.unassigned, j)
+	}
+}
+
+// renewLease renews the lease of the job with the given ID for the given
+// worker.
+func (s *jobQueue) renewLease(jobID, workerID string) error {
+	if jobID == "" {
+		return errors.New("jobID cannot be empty")
+	}
+	if workerID == "" {
+		return errors.New("workerID cannot be empty")
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	j, ok := s.jobs[jobID]
+	if !ok {
+		return errJobNotFound
+	}
+	if j.assignee != workerID {
+		return errJobNotAssigned
+	}
+
+	j.leaseExpiry = time.Now().Add(s.leaseTime)
+	return nil
+}
+
+// completeJob completes the job with the given ID for the given worker,
+// removing it from the jobQueue.
+func (s *jobQueue) completeJob(jobID, workerID string) error {
```

Comment:
In the context of the queue, "completing" a job may not make sense. We are technically just removing the job from the queue here.

```suggestion
func (s *jobQueue) removeJob(jobID, workerID string) error {
```

Reply:
True enough!

Reply:
Actually, it's a fairly specific operation where it only removes a job that is in-progress and assigned to the given worker. So, I think the intention of completeJob rings more true.
As with all things, I'll keep it in mind as the clay gets pushed around.